compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1054,22 +1054,107 @@ private boolean validateSpecifiedFields(BLangRecordLiteral mappingConstructor, B
         return isFieldsValid;
     }
 
-    private boolean validateRequiredFields(BRecordType type, List<RecordLiteralNode.RecordField> specifiedFields,
-                                           DiagnosticPos pos) {
+    private boolean validateRequiredAndReadonlyFields(BRecordType type,
+                                                      List<RecordLiteralNode.RecordField> specifiedFields,
+                                                      DiagnosticPos pos) {
         HashSet<String> specFieldNames = getFieldNames(specifiedFields);
         boolean hasAllRequiredFields = true;
+        boolean hasValidReadonlyFields = true;
 
         for (BField field : type.fields) {
-            // Check if `field` is explicitly assigned a value in the record literal
-            // If a required field is missing, it's a compile error
-            if (!specFieldNames.contains(field.name.value) && Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+            String fieldName = field.name.value;
+            BType fieldType = symTable.semanticError;
+            DiagnosticPos fieldPos;
+
+            if (specFieldNames.contains(fieldName)) {
+                RecordLiteralNode.RecordField recordField = getRecordFieldByName(specifiedFields, fieldName);
+                if (recordField == null) {
+                    continue;
+                }
+
+                switch (recordField.getKind()) {
+                    case RECORD_LITERAL_KEY_VALUE:
+                        BLangRecordKeyValueField keyValueField = (BLangRecordKeyValueField) recordField;
+                        fieldType = keyValueField.type;
+                        fieldPos = keyValueField.valueExpr.pos;
+                        break;
+                    case SIMPLE_VARIABLE_REF:
+                        BLangRecordLiteral.BLangRecordVarNameField varNameField =
+                                (BLangRecordLiteral.BLangRecordVarNameField) recordField;
+                        fieldType = varNameField.type;
+                        fieldPos = varNameField.pos;
+                        break;
+                    default:
+                        BLangRecordLiteral.BLangRecordSpreadOperatorField spreadOpField =
+                                (BLangRecordLiteral.BLangRecordSpreadOperatorField) recordField;
+                        BRecordType recordType = (BRecordType) spreadOpField.expr.type;
+
+                        for (BField bField : recordType.fields) {
+                            if (fieldName.equals(bField.name.value)) {
+                                fieldType = bField.type;
+                                break;
+                            }
+                        }
```

Comment:
Isn't this part duplicate of the logic in `getRecordFieldByName()`? Do we need to do the same here, after calling that method? Should be able to directly take the type of the spread field right?

Reply:
Yeah, noticed the duplication, but the issue was that we basically needed the type and pos, and for key-value and variable name fields we can retrieve a `RecordField` and get the type and pos from it. But for a spread-field the type is from a `BField` and the pos is still from the spread field. 

I've refactored this a bit with an inner class.